File: play-ahc-ws-standalone/src/main/java/play/libs/ws/ahc/StreamedResponse.java
/*
 * Copyright (C) from 2022 The Play Framework Contributors <https://github.com/playframework>, 2011-2021 Lightbend Inc. <https://www.lightbend.com>
 */

package play.libs.ws.ahc;

import org.apache.pekko.stream.javadsl.Source;
import org.apache.pekko.util.ByteString;
import org.reactivestreams.Publisher;
import play.api.libs.ws.ahc.AhcWSUtils;
import play.libs.ws.BodyReadable;
import play.libs.ws.StandaloneWSResponse;
import play.libs.ws.WSCookie;
import play.shaded.ahc.io.netty.handler.codec.http.HttpHeaderNames;
import play.shaded.ahc.org.asynchttpclient.HttpResponseBodyPart;
import scala.collection.Seq;
import scala.jdk.javaapi.StreamConverters;

import java.net.URI;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.TreeMap;
import java.util.function.Predicate;

import static java.util.stream.Collectors.toMap;
import scala.jdk.javaapi.CollectionConverters;

public class StreamedResponse implements StandaloneWSResponse, CookieBuilder {

    private final int status;
    private final Map<String, List<String>> headers;
    private final String statusText;
    private final URI uri;
    private final Publisher<HttpResponseBodyPart> publisher;
    private final StandaloneAhcWSClient client;
    private final boolean useLaxCookieEncoder;

    private List<WSCookie> cookies;

    public StreamedResponse(StandaloneAhcWSClient client,
                            int status,
                            String statusText, URI uri,
                            scala.collection.Map<String, Seq<String>> headers,
                            Publisher<HttpResponseBodyPart> publisher,
                            boolean useLaxCookieEncoder) {
        this.client = client;
        this.status = status;
        this.statusText = statusText;
        this.uri = uri;
        this.headers = asJava(headers);
        this.publisher = publisher;
        this.useLaxCookieEncoder = useLaxCookieEncoder;
    }

    @Override
    public int getStatus() {
        return status;
    }

    @Override
    public String getStatusText() {
        return statusText;
    }

    @Override
    public Map<String, List<String>> getHeaders() {
        return headers;
    }

    @Override
    public Object getUnderlying() {
        return publisher;
    }

    @Override
    public List<WSCookie> getCookies() {
        if (cookies == null) {
            cookies = buildCookies(headers);
        }
        return cookies;
    }

    @Override
    public Optional<WSCookie> getCookie(String name) {
        Predicate<WSCookie> predicate = (WSCookie c) -> c.getName().equals(name);
        return getCookies().stream().filter(predicate).findFirst();
    }

    @Override
    public String getContentType() {
        return getSingleHeader(HttpHeaderNames.CONTENT_TYPE.toString()).orElse("application/octet-stream");
    }

    @Override
    public <T> T getBody(BodyReadable<T> readable) {
        return readable.apply(this);
    }

    @Override
    public String getBody() {
        return getBodyAsBytes().decodeString(AhcWSUtils.getCharset(getContentType()));
    }

    @Override
    public ByteString getBodyAsBytes() {
        return client.blockingToByteString(getBodyAsSource());
    }

    @Override
    public Source<ByteString, ?> getBodyAsSource() {
        return Source.fromPublisher(publisher).map(bodyPart -> ByteString.fromArray(bodyPart.getBodyPartBytes()));
    }

    public boolean isUseLaxCookieEncoder() {
        return useLaxCookieEncoder;
    }

    @Override
    public URI getUri() {
        return this.uri;
    }

    private static java.util.Map<String, List<String>> asJava(scala.collection.Map<String, Seq<String>> scalaMap) {
        return StreamConverters.asJavaSeqStream(scalaMap).collect(toMap(f -> f._1(), f -> CollectionConverters.asJava(f._2()),
                (l, r) -> {
                    final List<String> merged = new ArrayList<>(l.size() + r.size());
                    merged.addAll(l);
                    merged.addAll(r);
                    return merged;
                },
                () -> new TreeMap<>(String.CASE_INSENSITIVE_ORDER)
            )
        );
    }

}

File: play-ahc-ws-standalone/src/main/scala/play/api/libs/ws/ahc/StreamedResponse.scala
/*
 * Copyright (C) from 2022 The Play Framework Contributors <https://github.com/playframework>, 2011-2021 Lightbend Inc. <https://www.lightbend.com>
 */

package play.api.libs.ws.ahc

import org.apache.pekko.stream.scaladsl.Source
import org.apache.pekko.util.ByteString
import org.reactivestreams.Publisher
import play.api.libs.ws.StandaloneWSResponse
import play.api.libs.ws.WSCookie
import play.shaded.ahc.org.asynchttpclient.HttpResponseBodyPart

import scala.collection.immutable.TreeMap
import scala.collection.mutable

/**
 * A streamed response containing a response header and a streamable body.
 *
 * Note that this is only usable with a stream call, i.e.
 *
 * {{{
 * import scala.concurrent.{ ExecutionContext, Future }
 *
 * import org.apache.pekko.util.ByteString
 * import org.apache.pekko.stream.scaladsl.Source
 *
 * import play.api.libs.ws.DefaultBodyReadables._
 * import play.api.libs.ws.ahc.StandaloneAhcWSClient
 *
 * class MyClass(ws: StandaloneAhcWSClient) {
 *   def doIt(implicit ec: ExecutionContext): Future[String] =
 *     ws.url("http://example.com").stream().map { response =>
 *        val _ = response.body[Source[ByteString, _]]
 *        ??? // process source to String
 *     }
 * }
 * }}}
 */
class StreamedResponse(
    client: StandaloneAhcWSClient,
    val status: Int,
    val statusText: String,
    val uri: java.net.URI,
    publisher: Publisher[HttpResponseBodyPart],
    val useLaxCookieEncoder: Boolean
) extends StandaloneWSResponse
    with CookieBuilder {

  def this(
      client: StandaloneAhcWSClient,
      status: Int,
      statusText: String,
      uri: java.net.URI,
      headers: Map[String, scala.collection.Seq[String]],
      publisher: Publisher[HttpResponseBodyPart],
      useLaxCookieEncoder: Boolean
  ) = {
    this(
      client,
      status,
      statusText,
      uri,
      publisher,
      useLaxCookieEncoder
    )
    origHeaders = headers
  }

  private var origHeaders: Map[String, scala.collection.Seq[String]] = Map.empty

  /**
   * Get the underlying response object.
   */
  override def underlying[T]: T = publisher.asInstanceOf[T]

  override lazy val headers: Map[String, scala.collection.Seq[String]] = {
    val mutableMap = mutable.TreeMap[String, scala.collection.Seq[String]]()(CaseInsensitiveOrdered)
    origHeaders.keys.foreach { name =>
      mutableMap.updateWith(name) {
        case Some(value) => Some(value ++ origHeaders.getOrElse(name, Seq.empty))
        case None        => Some(origHeaders.getOrElse(name, Seq.empty))
      }
    }
    TreeMap[String, scala.collection.Seq[String]]()(CaseInsensitiveOrdered) ++ mutableMap
  }

  /**
   * Get all the cookies.
   */
  override lazy val cookies: scala.collection.Seq[WSCookie] = buildCookies(headers)

  /**
   * Get only one cookie, using the cookie name.
   */
  override def cookie(name: String): Option[WSCookie] = cookies.find(_.name == name)

  /**
   * THIS IS A BLOCKING OPERATION. It should not be used in production.
   *
   * Note that this is not a charset aware operation, as the stream does not have access to the underlying machinery
   * that disambiguates responses.
   *
   * @return the body as a String
   */
  override lazy val body: String = bodyAsBytes.decodeString(AhcWSUtils.getCharset(contentType))

  /**
   * THIS IS A BLOCKING OPERATION. It should not be used in production.
   *
   * Note that this is not a charset aware operation, as the stream does not have access to the underlying machinery
   * that disambiguates responses.
   *
   * @return the body as a ByteString
   */
  override lazy val bodyAsBytes: ByteString = client.blockingToByteString(bodyAsSource)

  override lazy val bodyAsSource: Source[ByteString, ?] = {
    Source
      .fromPublisher(publisher)
      .map((bodyPart: HttpResponseBodyPart) => ByteString.fromArray(bodyPart.getBodyPartBytes))
  }

}
